### PR TITLE
Add "Enable enhanced measurement" toggle to Analytics Setup and Settings pages.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue:

- #7459 

## Relevant technical choices

There were a number of changes needed that weren't explicitly captured in the IB, I've tried to highlight them where useful to do so with comments on the PR or inline in the code.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/modules/analytics-4/components/common/EnhancedMeasurementSwitch.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+				{
+					'googlesitekit-analytics-enable-enhanced-measurement--loading':
+						loading,
+				}
+			) }
+		>
+			{ loading && (
+				<ProgressBar
+					small
+					className="googlesitekit-analytics-enable-enhanced-measurement__progress"
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/common/EnhancedMeasurementSwitch.js:78]: Note that trying to use the usual `height` prop for `ProgressBar` didn't give good results, hence the specific classname and corresponding styling here...

File: assets/js/modules/analytics-4/components/common/EnhancedMeasurementSwitch.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		)
+	);
+
+	const { setValues } = useDispatch( CORE_FORMS );
+
+	const handleClick = useCallback( () => {
+		setValues( ENHANCED_MEASUREMENT_FORM, {
+			[ ENHANCED_MEASUREMENT_ENABLED ]: ! isEnhancedMeasurementEnabled,
+		} );
+		onClick?.();
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/common/EnhancedMeasurementSwitch.js:62]: Although not in the IB, it was necessary to add a click handler here in order to be able to update the data stream `streamEnabled` setting in local state in `SettingsEnhancedMeasurementSwitch`.

File: assets/js/modules/analytics-4/components/common/PropertySelect.js [hunk 3]
```diff
@@ -78,21 +78,9 @@ export default function PropertySelect( props ) {
 			return false;
 		}
 
-		const isResolvingProperties =
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/common/PropertySelect.js:81]: It turns out we need to wait for both `PropertySelect` and `WebDataStreamSelect` to finish loading as part of the loading state logic in `SettingsEnhancedMeasurementSwitch`. To support that, I've extracted this loading logic into a reusable `isLoadingProperties()` selector.

File: assets/js/modules/analytics-4/components/common/WebDataStreamSelect.js [hunk 4]
```diff
@@ -75,38 +75,9 @@ export default function WebDataStreamSelect( props ) {
 			return false;
 		}
 
-		const loadedAccounts =
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/common/WebDataStreamSelect.js:78]: As per `PropertySelect`, this loading logic is extracted into a reusable `isLoadingWebDataStreams()` selector.

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+		return select( MODULES_ANALYTICS_4 ).isLoadingWebDataStreams( {
+			hasModuleAccess: hasAnalytics4Access,
+		} );
+	} );
+
+	const isEnhancedMeasurementStreamEnabled = useSelect( ( select ) => {
+		if (
+			! isValidPropertyID( propertyID ) ||
+			! isValidWebDataStreamID( webDataStreamID ) ||
+			isLoadingProperties ||
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js:84]: As mentioned above, we need to wait for the `PropertySelect` and `WebDataStreamSelect` loading states to settle, in part to synchronise the loading from a UI perspective, but also so we can avoid inadvertently making a request for the enhanced measurement settings for the wrong combination of `propertyID` and `webDataStreamID`.

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+			'getEnhancedMeasurementSettings',
+			[ propertyID, webDataStreamID ]
+		);
+	} );
+
+	const { setValues } = useDispatch( CORE_FORMS );
+	const { setEnhancedMeasurementStreamEnabled } =
+		useDispatch( MODULES_ANALYTICS_4 );
+
+	// If `isEnhancedMeasurementEnabled` is already defined in the first render, and either `PROPERTY_CREATE` or
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js:123]: This is required to support the cases where this component is rendered after the user has actioned the creation flow, i.e. "Set up a new property" or "Set up a new web data stream" is selected and they've pressed "Confirm Changes", the settings UI goes into the loading state, and then returns to the form view to show the "Additional permissions required" modal - _or_, when they continue to the OAuth flow and then show the form again upon return.

File: assets/js/modules/analytics-4/datastore/enhanced-measurement.js [hunk 7]
```diff
@@ -400,12 +406,21 @@ const baseSelectors = {
...
 	 */
-	isEnhancedMeasurementStreamEnabled( state, propertyID, webDataStreamID ) {
-		return !! state.enhancedMeasurement[ propertyID ]?.[ webDataStreamID ]
-			?.settings?.streamEnabled;
-	},
+	isEnhancedMeasurementStreamEnabled: createRegistrySelector(
+		( select ) => ( state, propertyID, webDataStreamID ) => {
+			const settings = select(
+				MODULES_ANALYTICS_4
+			).getEnhancedMeasurementSettings( propertyID, webDataStreamID );
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/enhanced-measurement.js:415]: This implementation detail was missed when the selector was first added in https://github.com/google/site-kit-wp/issues/7458, but is required.

File: assets/js/modules/analytics-4/datastore/enhanced-measurement.js [hunk 8]
```diff
@@ -427,6 +442,46 @@ const baseSelectors = {
...
+
+	/**
+	 * This is a utility selector that checks if either the GA4 module settings, or the
+	 * current web data stream's enhanced measurement settings have changed.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean} True if settings have changed, otherwise false.
+	 */
+	haveAnyGA4SettingsChanged: createRegistrySelector( ( select ) => () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/enhanced-measurement.js:454]: It was necessary to introduce this new utility selector as this dual check needs to be made in a number of places...

File: assets/js/modules/analytics/datastore/settings.js [hunk 9]
```diff
@@ -75,7 +75,7 @@ export const INVARIANT_INVALID_INTERNAL_PROPERTY_ID =
 	'cannot submit changes with incorrect internal webPropertyID';
 
 async function submitGA4Changes( { select, dispatch } ) {
-	if ( ! select( MODULES_ANALYTICS_4 ).haveSettingsChanged() ) {
+	if ( ! select( MODULES_ANALYTICS_4 ).haveAnyGA4SettingsChanged() ) {
```
Comment (techanvil) [inline assets/js/modules/analytics/datastore/settings.js:78]: It's necessary to add these changes to the `analytics` `submitChanges()` and `validateCanSubmitChanges()` functions, as it's the `analytics` versions of `submitChanges()` / `canSubmitChanges()` that are invoked on the Settings page.

File: assets/js/modules/analytics-4/components/common/WebDataStreamSelect.js (hunk 4 above)
Comment (tofumatt) [reply assets/js/modules/analytics-4/components/common/WebDataStreamSelect.js:78]: Nice 👍🏻

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+	} );
+
+	const isEnhancedMeasurementStreamEnabled = useSelect( ( select ) => {
+		if (
+			! isValidPropertyID( propertyID ) ||
+			! isValidWebDataStreamID( webDataStreamID ) ||
+			isLoadingProperties ||
+			isLoadingWebDataStreams
+		) {
+			return null;
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js:87]: It seems a bit odd to return `null` here given this is essentially a loading state, which we tend to return `undefined` for instead. 🤔

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.test.js [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+				);
+
+				if ( streamEnabled ) {
+					expect( switchControl ).toBeChecked();
+				} else {
+					expect( switchControl ).not.toBeChecked();
+				}
+			} );
+		}
+	);
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.test.js:149]: Seeing if statements in the tests makes them harder to read, as the reader has to backtrack above to make sense of the conditional logic here. I think this `.each` is saving minimal lines of code, but it's reducing the readability of these tests. In this case the test code isn't even shared because of this conditional.

I think it'd be better to write out each test spec without variables so the specs are written in plain English and the test code isn't conditional. Tests prioritising DRY principles reduce readability in exchange for conciseness, but for tests I think that's less useful.

File: assets/js/modules/analytics-4/components/setup/SetupEnhancedMeasurementSwitch.test.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+			accountID: '1000',
+		} );
+
+		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
+			propertyID: '2000',
+			webDataStreamID: '3000',
+		} );
+	} );
+
+	it( 'should render correctly, with the switch defaulting to the on position', () => {
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/components/setup/SetupEnhancedMeasurementSwitch.test.js:54]:
```suggestion
	it( 'should render with the switch defaulting to the on position', () => {
```

File: assets/js/modules/analytics-4/datastore/enhanced-measurement.js [hunk 13]
```diff
@@ -427,6 +442,46 @@ const baseSelectors = {
 
 		return ! isEqual( settings, savedSettings );
 	},
+
+	/**
+	 * This is a utility selector that checks if either the GA4 module settings, or the
+	 * current web data stream's enhanced measurement settings have changed.
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/datastore/enhanced-measurement.js:448]:
```suggestion
	 * This is a utility selector that checks if either the GA4 module settings, or the
	 * current web data stream's enhanced measurement settings, have changed.
```

File: assets/js/modules/analytics-4/datastore/enhanced-measurement.js [hunk 14]
```diff
@@ -427,6 +442,46 @@ const baseSelectors = {
...
+	haveAnyGA4SettingsChanged: createRegistrySelector( ( select ) => () => {
+		const {
+			getPropertyID,
+			getWebDataStreamID,
+			haveEnhancedMeasurementSettingsChanged,
+			haveSettingsChanged,
+		} = select( MODULES_ANALYTICS_4 );
+		if ( haveSettingsChanged() ) {
+			return true;
+		}
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/datastore/enhanced-measurement.js:463]: This might improve readability:

```suggestion
		} = select( MODULES_ANALYTICS_4 );

		if ( haveSettingsChanged() ) {
			return true;
		}
```

File: assets/js/modules/analytics-4/datastore/properties.test.js [hunk 15]
```diff
@@ -1347,5 +1349,157 @@ describe( 'modules/analytics-4 properties', () => {
...
+				registry
+					.dispatch( MODULES_ANALYTICS_4 )
+					.receiveGetWebDataStreamsBatch(
+						fixtures.webDataStreamsBatch,
+						{
+							propertyIDs: properties.map( ( { _id } ) => _id ),
+						}
+					);
+
+				// Verify that the selector returns false after the prelude.
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/datastore/properties.test.js:1482]: What does "prelude" here mean? I'm not quite clear on this comment 😅

**Review by tofumatt — CHANGES_REQUESTED** [review]
The code here looks really good and works well in usage.

I think there's some improvements to be made regarding test readability here, other than that not too many comments. Awesome stuff 👍🏻

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js (hunk 10 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js:87]: That's a good point - I've updated to return `undefined` here for the loading case. Please note, though, I've kept the `null` for the non-valid `propertyID` or `webDataStreamID` case, because this doesn't necessarily indicate a loading state per se, they can be in the `PROPERTY_CREATE` or `WEBDATASTREAM_CREATE` states and we don't want to return `undefined` in this case as `isEnhancedMeasurementStreamEnabled` being `undefined` value results in the `useEffect()` below being skipped - and we want it to run when `PROPERTY_CREATE` or `WEBDATASTREAM_CREATE` is selected.

File: assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.test.js (hunk 11 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.test.js:149]: I'm generally in full agreement here, preferring DAMP over DRY in a test context. As I was writing this test, I thought this particular case was "acceptable", due to the simple nature of the test. However I am more than happy to unroll these tests, and have done so. :)

File: assets/js/modules/analytics-4/datastore/enhanced-measurement.js (hunk 13 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/enhanced-measurement.js:448]: Ah, the old Oxford comma :)

I believe this is an optional grammatical construct, but I'm entirely happy to use it here :smile:

File: assets/js/modules/analytics-4/datastore/properties.test.js (hunk 15 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/properties.test.js:1482]: I just meant after the test case setup above it, but agree it's not the clearest. I've updated the comment, hopefully it's a bit more coherent now!

**Review by tofumatt — APPROVED** [review]
Awesome, looks great 😄
